File: src/algorithms/search/binary-search/binary-search.go
package binarysearch

// BinarySearch algorithm
func BinarySearch(array []int, target int) int {
	start := 0
	end := len(array)-1
	middle := (start+end)/2

	for start < end {

		expected := array[middle]
		if expected == target {
			return middle
		}

		if expected > target {
			end = middle -1
			middle = (start+end)/2
			continue
		}
		
		start = middle + 1
		middle = (start+end)/2

	}


	return -1}

File: algorithms/search/binary-search/binary-search.ts
export function binarySearch(array: number[], target: number): number {
  let start = 0
  let end = array.length - 1

  while (start <= end) {
    const middle = Math.floor((start + end) / 2)
    if (array[middle] === target) return middle

    if (target < array[middle]) {
      end = middle - 1
    }
    if (target > array[middle]) {
      start = middle + 1
    }
  }
  return -1
}
